### PR TITLE
Update training tricks docs

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Part of docs revamp #7666

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @akihironitta

### CODE REVIEW

File: docs/source/advanced/training_tricks.rst [hunk 1]
```diff
@@ -23,10 +28,40 @@ The effect is a large effective batch size of size KxN.
...
+    trainer = Trainer(accumulate_grad_batches=1)
+
+You can set different different values for it at different epochs by passing a dictionary, where key represent the epoch at which the value for gradient accumulation
+should be updated.
+
+.. testcode::
+
+        # from epoch 5, it starts accumulating every 2 batches. Here we have 4 instead of 5
+        # because epoch (key) should be zero-indexed.
+        trainer = Trainer(accumulate_grad_batches={4: 2})
```
Comment (tchaton) [inline docs/source/advanced/training_tricks.rst:44]:
```suggestion
        trainer = Trainer(accumulate_grad_batches={0: 8, 4: 4, 8: 1})
```

File: docs/source/advanced/training_tricks.rst [hunk 2]
```diff
@@ -5,16 +5,21 @@
 
 .. _training_tricks:
 
+###############
 Training Tricks
-================
-Lightning implements various tricks to help during training
+###############
+
+Lightning implements various tricks to help during training:
```
Comment (ananthsub) [inline docs/source/advanced/training_tricks.rst:12]: is there a different name we could use? as a user reading this, i'm nervous that i'm "tricked" by what lightning does, as opposed to learning about tools/techniques lightning offers to make my training smoother

the name was also based around the `TrainerTrainingTricksMixin` which has since been deprecated

File: docs/source/advanced/training_tricks.rst [hunk 3]
```diff
@@ -5,16 +5,21 @@
...
 
-Accumulate gradients
---------------------
-Accumulated gradients runs K small batches of size N before doing a backwards pass.
-The effect is a large effective batch size of size KxN.
+********************
+Accumulate Gradients
+********************
+
+Accumulated gradients runs K small batches of size N before doing a backwards pass. The effect is a large effective batch size of size KxN, where N is the batch size.
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply docs/source/advanced/training_tricks.rst:20]: the effective batch size should also mention the world size then, as this is a common confusion users ask about w/ gradient accumulation & distributed training

File: docs/source/advanced/training_tricks.rst [hunk 4]
```diff
@@ -5,16 +5,21 @@
...
+Lightning implements various tricks to help during training:
 
 ----------
 
-Accumulate gradients
---------------------
-Accumulated gradients runs K small batches of size N before doing a backwards pass.
-The effect is a large effective batch size of size KxN.
+********************
+Accumulate Gradients
```
Comment (ananthsub) [inline docs/source/advanced/training_tricks.rst:17]: should this be part of the optimization section?

Comment (rohitgr7) [reply docs/source/advanced/training_tricks.rst:17]: 100% will put it there and render it here as a direct copy.

File: docs/source/advanced/training_tricks.rst (hunk 2 above)
Comment (rohitgr7) [reply docs/source/advanced/training_tricks.rst:12]: needs to be something catchy (clickbait).
```
1. Training Tips
2. Effective Training Techniques
3. Effective Training Strategies
4. Training Strategies (maybe not to avoid conflicting with Strategy plugins but not sure)
5. Training Tactics
```
or something better you have in mind?

File: docs/source/advanced/training_tricks.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (rohitgr7) [reply docs/source/advanced/training_tricks.rst:20]: done!

File: docs/source/common/gradient_accumulation.rst [hunk 5]
```diff
@@ -0,0 +1,43 @@
+Accumulated gradients runs K small batches of size N before doing a backwards pass. The effect is a large effective batch size of size KxN, where N is the batch size.
+Internally it doesn't stack up the batches and do a forward pass, rather it accumulates the gradients for K batches and then do an ``optimizer.step`` to make sure the
+effective batch size is increased but there is no memory overhead.
+
+.. note::
```
Comment (tchaton) [inline docs/source/common/gradient_accumulation.rst:5]: warning

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by kaushikb11 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
